SamplePerformanceService: Refactor stats snapshot logic

ilya-bobyr opened the pull request:
#### Problem

Snapshot construction and interaction code was a bit more manual than necessary, even [causing a bug](https://github.com/solana-labs/solana/pull/30280) to slip past a review.

#### Summary of Changes

Separated snapshot construction from the diffing of two snapshots.
This should make the logic clearer.

Fixes #30312

steviez commented on core/src/sample_performance_service.rs:
```diff
@@ -51,19 +72,10 @@ impl SamplePerformanceService {
-            num_non_vote_transactions: bank.non_vote_transaction_count_since_restart(),
-            highest_slot,
-        };
+        let mut snapshot = Snapshot::from_forks(&bank_forks.read().unwrap());
```
```suggestion
        let mut snapshot = SamplePerformanceSnapshot::from_forks(&bank_forks.read().unwrap());
```

ilya-bobyr replied:
Sorry for sharing it half baked %)
Renamed `SamplePerformanceSnapshot` to `StatsSnapshot`.
It is already inside `sample_performance_service.rs`, so the `SamplePerformance` part in the name is just “noise”.

steviez commented on core/src/sample_performance_service.rs:
```diff
@@ -75,18 +87,10 @@ impl SamplePerformanceService {
-                    let bank_forks = bank_forks.read().unwrap();
-                    (bank_forks.root_bank(), bank_forks.highest_slot())
-                };
+                let new_snapshot = Snapshot::from_forks(&bank_forks.read().unwrap());
```
```suggestion
                let new_snapshot = SamplePerformanceSnapshot::from_forks(&bank_forks.read().unwrap());
```

CriesofCarrots commented:
I like the direction. Please give me a nudge when CI checks is happy

CriesofCarrots commented on core/src/sample_performance_service.rs:
```diff
@@ -75,18 +87,10 @@ impl SamplePerformanceService {
 
             if elapsed.as_secs() >= SAMPLE_INTERVAL {
                 now = Instant::now();
-                let (bank, highest_slot) = {
```
We do actually need the slot for writing to blockstore on L102

ilya-bobyr replied:
You are right.
I shared it a bit too quickly %)
Wanted to make sure you'll see it before you merge your PR.
Should be passing the CI now.

ilya-bobyr commented on core/src/sample_performance_service.rs:
```diff
@@ -116,3 +84,30 @@ impl SamplePerformanceService {
+        }
+    }
+
+    fn diff_since(&self, predecesor: &Self) -> (u64, u64, u64) {
```
Applied.

ilya-bobyr commented on core/src/sample_performance_service.rs:
```diff
@@ -51,19 +45,7 @@ impl SamplePerformanceService {
-            num_non_vote_transactions: bank.non_vote_transaction_count_since_restart(),
-            highest_slot,
-        };
+        let mut snapshot = StatsSnapshot::from_forks(&bank_forks.read().unwrap());
```
I figured, it makes the `from_forks` argument type a bit simpler.
Otherwise, I though, would need to pass `&Arc<RwLock<BankForks>>`.
And I'm used to treating `Arc`s as value types, rather than passing them by reference.
But in my set of rules, I figured, a function should take an `Arc` only when it will actually need to use that `Arc`.
If it does not, it is more generic to use a reference.
Just realized, I can make it `&RwLock<BankForks>`.
This way it shows that `from_forks` will lock the argument.

CriesofCarrots commented on core/src/sample_performance_service.rs:
```diff
@@ -75,18 +57,12 @@ impl SamplePerformanceService {
-                    .saturating_sub(snapshot.num_non_vote_transactions);
+                let (num_transactions, num_non_vote_transactions, num_slots) =
+                    new_snapshot.diff_since(&snapshot);
+                snapshot = new_snapshot;
```
Could you add a comment about this line? It's just easy to miss when scanning the code. Maybe also set off with line breaks?
Could steal Steve's verbiage, something like:
Store the new snapshot to compare against in the next iteration of the loop

ilya-bobyr replied:
Added a comment and an empty line to create a paragraph.

CriesofCarrots reviewed (dismissed):
Nice, code lgtm! Just one nit to make the snapshot re-assignment more obvious

steviez reviewed (approved):
Looks good to me!

The only thing I'm on the fence about is whether we retain a comment that mentions that our sampling of root bank `Bank` (in `from_forks()`) returns "all-time" transaction counts.

But maybe that kind of comment would be better off living with the functions in `bank.rs`, and I'm just over-fixating on this aspect since we all missed it in review

CriesofCarrots commented:
https://github.com/solana-labs/solana/issues/30312#issuecomment-1430290075

ilya-bobyr commented:
> Looks good to me!
> 
> The only thing I'm on the fence about is whether we retain a comment that mentions that our sampling of root bank `Bank` (in `from_forks()`) returns "all-time" transaction counts.
> 
> But maybe that kind of comment would be better off living with the functions in `bank.rs`, and I'm just over-fixating on this aspect since we all missed it in review

The reason I removed it is that there is a clear separation now.
`StatsSnapshot` is recording the absolute count, and `diff_since` returns the difference.
I figured, written like this, the code is clear enough to have no need for the comment.

If we want to be even more explicit, I though, one option is to use different names for the variables that hold the differences:

```rust
                let (sample_num_transactions, sample_num_non_vote_transactions, sample_num_slots) =
                    new_snapshot.diff_since(&snapshot);
                // Store the new snapshot to compare against in the next iteration of the loop.
                snapshot = new_snapshot;

                let highest_slot = snapshot.highest_slot;

                let perf_sample = PerfSampleV2 {
                    num_slots: sample_num_slots,
                    num_transactions: sample_num_transactions,
                    num_non_vote_transactions: sample_num_non_vote_transactions,
                    sample_period_secs: elapsed.as_secs() as u16,
                };
```

But it looked too long.  And the function is so short now, that it seems readable enough.
At the same time, considering I've rewritten it 2 or 3 times now, I might be biased.

steviez reviewed (dismissed):
Looks good to me, pending possibly shipping https://github.com/solana-labs/solana/pull/30316 first and then rebasing on top

ilya-bobyr commented:
Rebased on top of #30316.

CriesofCarrots reviewed (approved):
Awesome, lgtm! And fwiw, I agree with you about the comments no longer being necessary.

steviez reviewed (approved):
> StatsSnapshot is recording the absolute count, and diff_since returns the difference.
I figured, written like this, the code is clear enough to have no need for the comment.

I didn't explicitly state it, but my last ship-it implied that I agreed with you on this aspect.
